api/src/exam-environment/routes/exam-environment.ts

```diff
@@ -0,0 +1,376 @@
+  const generatedExamContent = generateExam(exam);
+
+  const maybeGeneratedExam = await mapErr(
+    this.prisma.generatedExam.create({
+      data: generatedExamContent
+    })
+  );
+
+  if (maybeGeneratedExam.error !== null) {
+    void reply.code(500);
+    return reply.send(
+      // TODO: Consider more specific code
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(
+        'Unable to generate exam, due to: ' +
+          JSON.stringify(maybeGeneratedExam.error)
+      )
+    );
+  }
+
+  const generatedExam = maybeGeneratedExam.data;
+
+  // Create exam attempt so, even if user disconnects, their attempt is still recorded:
+  const { error, data: examAttempt } = await mapErr(
+    this.prisma.newExamAttempt.create({
+      data: {
+        user_id: user.id,
+        exam_id: exam.id,
+        generated_exam_id: generatedExam.id,
+        start_time: Date.now(),
+        question_types: [],
+        needs_retake: false
+      }
+    })
+  );
+
+  if (error !== null) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT_CREATE_EXAM_ATTEMPT(JSON.stringify(error))
+    );
+  }
+  // NOTE: Anything that goes wrong after this point needs to unwind the exam attempt.
+
+  const maybeUserExam = syncMapErr(() => createUserExam(generatedExam, exam));
+
+  if (maybeUserExam.error !== null) {
+    await this.prisma.newExamAttempt.delete({
+      where: {
+        id: examAttempt.id
+      }
+    });
+    await this.prisma.generatedExam.delete({
+      where: {
+        id: generatedExam.id
+      }
+    });
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeUserExam.error))
+    );
+  }
+
+  const userExam = maybeUserExam.data;
+
+  void reply.code(200);
+  return reply.send({
+    data: {
+      exam: userExam,
+      examAttempt
+    }
+  });
+}
+
+/**
+ * Handles the submission of an exam attempt.
+ *
+ * Requires token to be validated.
+ *
+ * TODO: Consider validating req.user.id == attempt.user_id?
+ */
+async function postExamAttemptHandler(
+  this: FastifyInstance,
+  req: UpdateReqType<typeof schemas.examEnvironmentPostExamAttempt>,
+  reply: FastifyReply
+) {
+  const { attempt } = req.body;
+
+  // Get generated exam from database
+  const maybeGeneratedExam = await mapErr(
+    this.prisma.generatedExam.findUnique({
+      where: {
+        id: attempt.generated_exam_id
+      }
+    })
+  );
+
+  if (maybeGeneratedExam.error !== null) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeGeneratedExam.error))
+    );
+  }
+
+  const generatedExam = maybeGeneratedExam.data;
+
+  if (generatedExam === null) {
+    void reply.code(404);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT('Generated exam not found.')
+    );
+  }
+  // Ensure attempt matches generated exam
+  const maybeValidExamAttempt = syncMapErr(() =>
+    validateAttempt(generatedExam, attempt)
+  );
+
```

Comment:
Don't trust the attemptId coming from the user. Get the attemptId from the database (userId).

Or make sure the id that comes in is their latest attempt.